src/vm_optimizations.nit

```diff
@@ -0,0 +1,353 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Copyright 2015 Julien Pagès <julien.pages@lirmm.fr>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Optimization of the nitvm
+module vm_optimizations
+
+import vm
+
+redef class VirtualMachine
+
+	# Add optimization of the method dispatch
+	redef fun callsite(callsite: nullable CallSite, arguments: Array[Instance]): nullable Instance
+	do
+		var initializers = callsite.mpropdef.initializers
+		if not initializers.is_empty then
+			var recv = arguments.first
+			var i = 1
+			for p in initializers do
+				if p isa MMethod then
+					var args = [recv]
+					for x in p.intro.msignature.mparameters do
+						args.add arguments[i]
+						i += 1
+					end
+					self.send(p, args)
+				else if p isa MAttribute then
+					assert recv isa MutableInstance
+					write_attribute(p, recv, arguments[i])
+					i += 1
+				else abort
+			end
+			assert i == arguments.length
+
+			return send_optimize(callsite.as(not null), [recv])
+		end
+
+		return send_optimize(callsite.as(not null), arguments)
+	end
+
+	# Try to have the most efficient implementation of the method dispatch
+	fun send_optimize(callsite: CallSite, args: Array[Instance]): nullable Instance
+	do
+		var recv = args.first
+		var mtype = recv.mtype
+		var ret = send_commons(callsite.mproperty, args, mtype)
+		if ret != null then return ret
+
+		if callsite.status == 0 then callsite.optimize(recv)
+
+		var propdef
+		if callsite.status == 1 then
+			propdef = method_dispatch_sst(recv.vtable.internal_vtable, callsite.offset)
+		else
+			propdef = method_dispatch_ph(recv.vtable.internal_vtable, recv.vtable.mask,
+				callsite.id, callsite.offset)
+		end
+
+		return self.call(propdef, args)
+	end
+end
+
+redef class AAttrExpr
+	# The relative position of this attribute if perfect hashing is used,
+	# The absolute position of this attribute if SST is used
+	var offset: Int
+
+	# 0: default value
```

Comment:
You should be explaining what is a `AAttrExpr::status` before giving the possible values.